File: marvin_masterclass.py
import webbrowser
import os
import requests
import sqlite3
import sys
import subprocess

# implement pip as a subprocess to install pandas and requests package if it is not present in the host system
subprocess.check_call([sys.executable, '-m', 'pip', 'install', 'pandas'])
subprocess.check_call([sys.executable, '-m', 'pip', 'install', 'requests'])
import pandas as pd
pd.set_option("display.max_rows", None,'display.width', None)

print("Fetching data.............. This will take some time(less than a minute) kindly wait...............")
data = requests.get('https://www.datim.org/api/sqlViews/fgUtV6e9YIX/data.json').json()
print("Data fetched successfully....................")

#Extracting the column labels into a list of tuple
gridList = data['listGrid']['headers']
dataColumns = []

for i in range(0,len(gridList)):
    dataColumns.append(gridList[i]['column'])

columnTuple = tuple(dataColumns)


#Extracting the rows data into an array of tuples
dataRows = data['listGrid']['rows']
rowTuple = []

 
for row in range(0,len(dataRows)):
    rowTuple.append(tuple(dataRows[row])) 


"""Creation of SQLite3 database """
if (os.path.isfile("mbathamarvin.db")):
    connection = sqlite3.connect("mbathamarvin.db")
    cursor = connection.cursor()
else:
    connection = sqlite3.connect("mbathamarvin.db")
    cursor = connection.cursor()
    cursor.execute("""create table masterclass ( mechanism text, code text,uid text,
    partner text, primeid text, agency text, ou text,startdate text,enddate text)""")
    connection.commit()

cursor.executemany("insert into masterclass values (?,?,?,?,?,?,?,?,?)", rowTuple )

#quering the database and storing response in a pandas dataframe
query_df = pd.read_sql_query("select * from masterclass", connection)

connection.close()

#creating the html file, saving it to the curent directory and displaying content in the default browser
query_df.to_html("Table.html")

if (os.path.isfile("Table.html")):
    query_df.to_html("Table.html")
    print("Table.html created successfuly")
    filename = 'file:///'+os.getcwd()+'/' + 'Table.html'
    webbrowser.open_new_tab(filename)
else:
    filename = 'file:///'+os.getcwd()+'/' + 'Table.html'
    webbrowser.open_new_tab(filename)

File: marvin_masterclass.js
const https = require('https');
const sqlite3 = require('sqlite3');
const fs = require('fs');
const { config } = require('process');

console.log('Fetching Data ..................Kindly be patient............................')

let url = "https://www.datim.org/api/sqlViews/fgUtV6e9YIX/data.json";


let json;

https.get(url,(res) => {
    let body = "";

    res.on("data", (chunk) => {
        body += chunk;
    });

    res.on("end", () => {
        try {
            json = JSON.parse(body);
            console.log('Data recieved .............Parsing the data..................');
            

        } catch (error) {
            console.error(error.message);
        };
    });

}).on("error", (error) => {
    console.error(error.message);
});



setTimeout(() => {

    const myFirst = json.listGrid.headers
    console.log('A list of column names: ')

    //getting the columns names
    let columnNames = []
    for (let index = 0; index < myFirst.length; index++){
        // console.log(myFirst[index].name)
        columnNames.push(myFirst[index].name)
    } 
    console.log(columnNames)
    // console.log(Object.keys(myFirst))


    //getting the rows data
    console.log('Retrieving the rows data from thr json object.....')
    RowsData = json.listGrid.rows 


var db;

//create the database if it does not exist and populate it 
if (!fs.existsSync('myFirst.db')){
    db = new sqlite3.Database('./myFirst.db',(err) => {
        if (err){
            console.log(err)
        }else{
            console.log('Database created Successfully')
        }
    })
    
    setTimeout( () => {
        db.run(`create table masterclass (mechanism text, code text, uid text, partner text,  primeid text, agency text, ou text, startdate text, enddate text)`);
        console.log('Created the table successfully')
    }, 300);
    
    //waiting for the database table to be created before inserting data
    setTimeout( () => {

        for (let record = 0; record < RowsData.length; record++){
            db.run(`
            insert into masterclass (mechanism, code, uid, partner, primeid, agency, ou, startdate, enddate) values (?, ?, ?, ?, ?, ?, ?, ?, ?);`, 
                [RowsData[record][0],RowsData[record][1],RowsData[record][2]
                ,RowsData[record][3],RowsData[record][4],RowsData[record][5],
                RowsData[record][6],RowsData[record][7],RowsData[record][8]], 
            (err) => {
                    if (err){
                        console.log(err)
                    }else {
                        console.log('succesfully inserted')
                    }
                })
        }
    }, 1500);
    
}else{
    db = new sqlite3.Database('./myFirst.db',(err) => {
        if (err){
            console.log(err)
        }else{
            console.log('Database connected Successfully')
        }
    })
    console.log('The database already exist')
}

let queryData = []
setTimeout(() => {

    db.all('select * from masterclass',[], (err, rows) => {
        if (err){
            console.log(err);
        }else{
            rows.forEach((row) =>{
            queryData.push(row)
            // console.log(row)
            })
        }
    //console.log(queryData)
    })
}, 2000);



setTimeout(() => {
    console.log('The data below is from the database query .............')
    console.log(queryData)
    console.log('Total query item ' + queryData.length)
    console.log('End of data query data.............................')
}, 5000);

// setTimeout(() => {
//     db.close()
// }, 20000)



}, 95000);

